lib/nimble_options.ex

```diff
@@ -558,8 +567,10 @@ defmodule NimbleOptions do
   end
 
   defp validate_option({opts, orig_opts}, key, schema) do
-    with {:ok, value} <- validate_value({opts, orig_opts}, key, schema),
-         {:ok, value} <- validate_type(schema[:type], key, value) do
+    with {:ok, value} <-
+           validate_value({opts, orig_opts}, key, schema),
+         {:ok, value} <-
+           validate_type(schema[:type], key, value, Keyword.get(schema, :redact, false)) do
```

Comment:
@whatyouhide I'm not sure if this is the right approach here. I kept running into scenarios where `:redact` wasn't available on the schema. When I ran into issues it seemed like it was always when `key` was a schema field. Like `:doc`. I'm not sure if I'm doing something incorrectly or this is a quirk of some sort.

Reply:
I think it might be something wrong somewhere because if `:redact` has a default it _should_ be here.

Reply:
That's is what I thought too. I spent almost a whole day debugging this but haven't figured it out yet. I don't think I'm doing anything that would affect how that works.

Reply:
It seems to be happening on on `:doc` and `:type_doc`. Both of which have a type of `{:or, subtype}`. Anything weird about that type when validating the schema?

Here is my test output. Hopefully that's helpful.

```
1) test NimbleOptions.docs/1 uses the type_doc option (NimbleOptions.DocsTest)
     test/nimble_options/docs_test.exs:98
     ** (FunctionClauseError) no function clause matching in NimbleOptions.error_tuple/5

     The following arguments were given to NimbleOptions.error_tuple/5:

         # 1
         :type_doc

         # 2
         false

         # 3
         "invalid value for :type_doc option: expected string"

         # 4
         ", got: false"

         # 5
         nil

     Attempted function clauses (showing 2 out of 2):

         defp error_tuple(key, value, message, _, true)
         defp error_tuple(key, value, message, unredacted_message, false)

     code: assert NimbleOptions.docs(NimbleOptions.new!(schema)) == docs
     stacktrace:
       (nimble_options 1.1.1) lib/nimble_options.ex:1154: NimbleOptions.error_tuple/5
       (nimble_options 1.1.1) lib/nimble_options.ex:886: anonymous fn/5 in NimbleOptions.validate_type/4
       (elixir 1.17.2) lib/enum.ex:4858: Enumerable.List.reduce/3
       (elixir 1.17.2) lib/enum.ex:2585: Enum.reduce_while/3
       (nimble_options 1.1.1) lib/nimble_options.ex:876: NimbleOptions.validate_type/4
       (nimble_options 1.1.1) lib/nimble_options.ex:573: NimbleOptions.validate_option/3
       (nimble_options 1.1.1) lib/nimble_options.ex:551: NimbleOptions.reduce_options/2
       (elixir 1.17.2) lib/enum.ex:4858: Enumerable.List.reduce/3
       (elixir 1.17.2) lib/enum.ex:2585: Enum.reduce_while/3
       (nimble_options 1.1.1) lib/nimble_options.ex:544: NimbleOptions.validate_options/2
       (nimble_options 1.1.1) lib/nimble_options.ex:517: NimbleOptions.validate_options_with_schema_and_path/3
       (nimble_options 1.1.1) lib/nimble_options.ex:551: NimbleOptions.reduce_options/2
       (elixir 1.17.2) lib/enum.ex:4858: Enumerable.List.reduce/3
       (elixir 1.17.2) lib/enum.ex:2585: Enum.reduce_while/3
       (nimble_options 1.1.1) lib/nimble_options.ex:544: NimbleOptions.validate_options/2
       (nimble_options 1.1.1) lib/nimble_options.ex:517: NimbleOptions.validate_options_with_schema_and_path/3
       (nimble_options 1.1.1) lib/nimble_options.ex:379: NimbleOptions.new!/1
       test/nimble_options/docs_test.exs:116: (test)

...............................................................

  2) test type validation invalid {:or, subtypes} with nested keyword lists (NimbleOptionsTest)
     test/nimble_options_test.exs:1060
     ** (FunctionClauseError) no function clause matching in NimbleOptions.error_tuple/5

     The following arguments were given to NimbleOptions.error_tuple/5:

         # 1
         :enabled

         # 2
         "not a boolean"

         # 3
         "invalid value for :enabled option: expected boolean"

         # 4
         ", got: \"not a boolean\""

         # 5
         nil

     Attempted function clauses (showing 2 out of 2):

         defp error_tuple(key, value, message, _, true)
         defp error_tuple(key, value, message, unredacted_message, false)

     code: assert NimbleOptions.validate(opts, schema) == {
     stacktrace:
       (nimble_options 1.1.1) lib/nimble_options.ex:1154: NimbleOptions.error_tuple/5
       (nimble_options 1.1.1) lib/nimble_options.ex:573: NimbleOptions.validate_option/3
       (nimble_options 1.1.1) lib/nimble_options.ex:551: NimbleOptions.reduce_options/2
       (elixir 1.17.2) lib/enum.ex:4858: Enumerable.List.reduce/3
       (elixir 1.17.2) lib/enum.ex:2585: Enum.reduce_while/3
       (nimble_options 1.1.1) lib/nimble_options.ex:544: NimbleOptions.validate_options/2
       (nimble_options 1.1.1) lib/nimble_options.ex:517: NimbleOptions.validate_options_with_schema_and_path/3
       (nimble_options 1.1.1) lib/nimble_options.ex:888: anonymous fn/5 in NimbleOptions.validate_type/4
       (elixir 1.17.2) lib/enum.ex:4858: Enumerable.List.reduce/3
       (elixir 1.17.2) lib/enum.ex:2585: Enum.reduce_while/3
       (nimble_options 1.1.1) lib/nimble_options.ex:876: NimbleOptions.validate_type/4
       (nimble_options 1.1.1) lib/nimble_options.ex:573: NimbleOptions.validate_option/3
       (nimble_options 1.1.1) lib/nimble_options.ex:551: NimbleOptions.reduce_options/2
       (elixir 1.17.2) lib/enum.ex:4858: Enumerable.List.reduce/3
       (elixir 1.17.2) lib/enum.ex:2585: Enum.reduce_while/3
       (nimble_options 1.1.1) lib/nimble_options.ex:544: NimbleOptions.validate_options/2
       (nimble_options 1.1.1) lib/nimble_options.ex:517: NimbleOptions.validate_options_with_schema_and_path/3
       test/nimble_options_test.exs:1096: (test)

...................

  3) test :doc valid documentation for key (NimbleOptionsTest)
     test/nimble_options_test.exs:189
     ** (FunctionClauseError) no function clause matching in NimbleOptions.error_tuple/5

     The following arguments were given to NimbleOptions.error_tuple/5:

         # 1
         :doc

         # 2
         false

         # 3
         "invalid value for :doc option: expected string"

         # 4
         ", got: false"

         # 5
         nil

     Attempted function clauses (showing 2 out of 2):

         defp error_tuple(key, value, message, _, true)
         defp error_tuple(key, value, message, unredacted_message, false)

     code: assert NimbleOptions.validate([], schema) == {:ok, [context: 1]}
     stacktrace:
       (nimble_options 1.1.1) lib/nimble_options.ex:1154: NimbleOptions.error_tuple/5
       (nimble_options 1.1.1) lib/nimble_options.ex:886: anonymous fn/5 in NimbleOptions.validate_type/4
       (elixir 1.17.2) lib/enum.ex:4858: Enumerable.List.reduce/3
       (elixir 1.17.2) lib/enum.ex:2585: Enum.reduce_while/3
       (nimble_options 1.1.1) lib/nimble_options.ex:876: NimbleOptions.validate_type/4
       (nimble_options 1.1.1) lib/nimble_options.ex:573: NimbleOptions.validate_option/3
       (nimble_options 1.1.1) lib/nimble_options.ex:551: NimbleOptions.reduce_options/2
       (elixir 1.17.2) lib/enum.ex:4858: Enumerable.List.reduce/3
       (elixir 1.17.2) lib/enum.ex:2585: Enum.reduce_while/3
       (nimble_options 1.1.1) lib/nimble_options.ex:544: NimbleOptions.validate_options/2
       (nimble_options 1.1.1) lib/nimble_options.ex:517: NimbleOptions.validate_options_with_schema_and_path/3
       (nimble_options 1.1.1) lib/nimble_options.ex:551: NimbleOptions.reduce_options/2
       (elixir 1.17.2) lib/enum.ex:4858: Enumerable.List.reduce/3
       (elixir 1.17.2) lib/enum.ex:2585: Enum.reduce_while/3
       (nimble_options 1.1.1) lib/nimble_options.ex:544: NimbleOptions.validate_options/2
       (nimble_options 1.1.1) lib/nimble_options.ex:517: NimbleOptions.validate_options_with_schema_and_path/3
       (nimble_options 1.1.1) lib/nimble_options.ex:379: NimbleOptions.new!/1
       (nimble_options 1.1.1) lib/nimble_options.ex:355: NimbleOptions.validate/2
       test/nimble_options_test.exs:193: (test)

..........................................

  4) test :doc invalid documentation for key (NimbleOptionsTest)
     test/nimble_options_test.exs:196
     Expected exception ArgumentError but got FunctionClauseError (no function clause matching in NimbleOptions.error_tuple/5)
     stacktrace:
       (nimble_options 1.1.1) lib/nimble_options.ex:1154: NimbleOptions.error_tuple(:doc, 1, "invalid value for :doc option: expected string", ", got: 1", nil)
       (nimble_options 1.1.1) lib/nimble_options.ex:886: anonymous fn/5 in NimbleOptions.validate_type/4
       (elixir 1.17.2) lib/enum.ex:4858: Enumerable.List.reduce/3
       (elixir 1.17.2) lib/enum.ex:2585: Enum.reduce_while/3
       (nimble_options 1.1.1) lib/nimble_options.ex:876: NimbleOptions.validate_type/4
       (nimble_options 1.1.1) lib/nimble_options.ex:573: NimbleOptions.validate_option/3
       (nimble_options 1.1.1) lib/nimble_options.ex:551: NimbleOptions.reduce_options/2
       (elixir 1.17.2) lib/enum.ex:4858: Enumerable.List.reduce/3
       (elixir 1.17.2) lib/enum.ex:2585: Enum.reduce_while/3
       (nimble_options 1.1.1) lib/nimble_options.ex:544: NimbleOptions.validate_options/2
       (nimble_options 1.1.1) lib/nimble_options.ex:517: NimbleOptions.validate_options_with_schema_and_path/3
       (nimble_options 1.1.1) lib/nimble_options.ex:551: NimbleOptions.reduce_options/2
       (elixir 1.17.2) lib/enum.ex:4858: Enumerable.List.reduce/3
       (elixir 1.17.2) lib/enum.ex:2585: Enum.reduce_while/3
       (nimble_options 1.1.1) lib/nimble_options.ex:544: NimbleOptions.validate_options/2
       (nimble_options 1.1.1) lib/nimble_options.ex:517: NimbleOptions.validate_options_with_schema_and_path/3
       (nimble_options 1.1.1) lib/nimble_options.ex:379: NimbleOptions.new!/1
       (nimble_options 1.1.1) lib/nimble_options.ex:355: NimbleOptions.validate/2
```

Reply:
The more I dig into this the more I think what I'm currently doing is correct. `NimbleOptions.new!/1` is responsible for validating the schema. The schema used to validate the schema is kept static while it goes through and builds out the "validated schema". The validated schema that is returned by `NimbleOptions.new!/1` will have schema options that include defaults. But the schema that is being used to validate that schema is static so it doesn't include any default schema options. This is where we run into this problem.

The only way to fix this with how things currently work would be to do something like I'm currently doing or add `:redact` to all keys in [`@options_schema`](https://github.com/dashbitco/nimble_options/blob/cc1bd8161d000e7430390cb09510687ae445e6ef/lib/nimble_options.ex#L2-L91).

Sorry for the confusing language. It's all very circular which is why I've been having a hard time debugging this.

Reply:
There's also an issue with the `{:or, subtype}` type. When using `:keyword_list`, `:non_empty_keyword_list` and `:map` the keys are used as a schema but never get converted into a schema. So `:redact` isn't present and we see the same issue.

For instance, using this schema

```
[
  docs: [
    type: {:or, [:boolean, keyword_list: [enabled: [type: :boolean]]]}
  ]
]
```

the subtype `[enabled: [type: :boolean]]` is used as a schema but isn't converted into a schema via `NimbleOptions.new!/1`. The conversation process is what adds schema options with a default value.

I'm not sure if this is fine or a larger issue with `{:or, subtype}`. I'm happy to open up another issue for this if you think that makes sense.

Either way, it seems like there are currently multiple use-cases where it's reasonable to expect a schema option to be `nil` despite it having a default value.

Reply:
Okay I see. For now it's ok, we can keep checking for `:redact` when validating the schema, I'll see if I can come up with something locally after we merge this 🙃

Reply:
Sounds good! Thank you for working through this with me. I’ve looked at solutions for what I talked about above so if you’d like help let me know.